prov/rxm/src/rxm_ep.c

```diff
@@ -1141,160 +1146,121 @@ rxm_ep_sar_tx_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+	inject_pkt->ctrl_hdr.conn_id = rxm_conn->handle.remote_key;
+	memcpy(inject_pkt->data, buf, len);
 
-	if (pkt_size <= rxm_ep->msg_info->tx_attr->inject_size) {
-		inject_pkt->hdr.size = len;
-		inject_pkt->hdr.tag = tag;
-		inject_pkt->ctrl_hdr.conn_id = rxm_conn->handle.remote_key;
-		memcpy(inject_pkt->data, buf, len);
-		return rxm_ep_inject_send(rxm_ep, rxm_conn, inject_pkt, pkt_size);
-	} else {
-		return rxm_ep_emulate_inject(rxm_ep, rxm_conn, buf, len, pkt_size,
-					     0, flags, tag, inject_pkt->hdr.op);
-	}
+	return rxm_ep_msg_inject_send(rxm_ep, rxm_conn, inject_pkt,
+				      len + sizeof(struct rxm_pkt),
+				      rxm_ep->util_ep.tx_cntr_inc);
 }
 
 static inline ssize_t
-rxm_ep_inject_common(struct rxm_ep *rxm_ep, const void *buf, size_t len,
-		     fi_addr_t dest_addr, uint64_t data, uint64_t flags,
-		     uint64_t tag, uint8_t op)
+rxm_ep_inject_send(struct rxm_ep *rxm_ep, const void *buf, size_t len,
+		   fi_addr_t dest_addr, uint64_t data, uint64_t flags,
+		   uint64_t tag, uint8_t op)
 {
 	struct rxm_conn *rxm_conn;
-	size_t pkt_size = sizeof(struct rxm_pkt) + len;
+	struct rxm_tx_base_buf *tx_buf;
 	ssize_t ret;
 
-	assert(len <= rxm_ep->rxm_info->tx_attr->inject_size);
+	assert(len <= rxm_ep_get_inject_buf_limit(rxm_ep));
 
-	ret = rxm_acquire_conn_connect(rxm_ep, dest_addr, &rxm_conn);
+	ret = rxm_ep_send_prerequisites(rxm_ep, dest_addr, &rxm_conn);
 	if (OFI_UNLIKELY(ret))
 		return ret;
 
-	if (OFI_UNLIKELY(!dlist_empty(&rxm_conn->deferred_tx_queue))) {
-		rxm_ep_progress_multi(&rxm_ep->util_ep);
-		if (!dlist_empty(&rxm_conn->deferred_tx_queue))
-			return -FI_EAGAIN;
+	tx_buf = (struct rxm_tx_base_buf *)
+		rxm_tx_buf_get(rxm_ep, RXM_BUF_POOL_TX_INJECT);
+	if (OFI_UNLIKELY(!tx_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+			"Ran out of buffers from Eager Inject buffer pool\n");
+		return -FI_EAGAIN;
 	}
 
-	if (pkt_size <= rxm_ep->msg_info->tx_attr->inject_size) {
-		struct rxm_tx_base_buf *tx_buf = (struct rxm_tx_base_buf *)
-			rxm_tx_buf_get(rxm_ep, RXM_BUF_POOL_TX_INJECT);
-		if (OFI_UNLIKELY(!tx_buf)) {
-			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
-				"Ran out of buffers from Eager Inject buffer pool\n");
-			return -FI_EAGAIN;
-		}
-		rxm_ep_format_tx_buf_pkt(rxm_conn, len, op, data, tag,
-					 flags, &tx_buf->pkt);
-		memcpy(tx_buf->pkt.data, buf, len);
-		ret = rxm_ep_inject_send(rxm_ep, rxm_conn, &tx_buf->pkt, pkt_size);
-		/* release allocated buffer for further reuse */
-		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX_INJECT, tx_buf);
-		return ret;
-	} else {
-		return rxm_ep_emulate_inject(rxm_ep, rxm_conn, buf, len,
-					     pkt_size, data, flags, tag, op);
-	}
+	rxm_ep_format_tx_buf_pkt(rxm_conn, len, op, data, tag,
+				 flags, &tx_buf->pkt);
+	memcpy(tx_buf->pkt.data, buf, len);
+
+	ret = rxm_ep_msg_inject_send(rxm_ep, rxm_conn, &tx_buf->pkt,
+				     len + sizeof(struct rxm_pkt),
+				     rxm_ep->util_ep.tx_cntr_inc);
+
+	/* release allocated buffer for further reuse */
```

Comment:
nit: as long as we're changing the code here, let's drop the unnecessary comment.